omegatrade-app/frontend/src/app/components/simulation/simulation.component.ts

```diff
@@ -0,0 +1,148 @@
+  }
+
+  ngOnInit(): void {
+    this.getCompanies();
+    this.simulateForm = this.formBuilder.group({
+      companyId: ['', [Validators.required]],
+      timeInterval: ['', [Validators.required]],
+      data: ['', [Validators.required]],
+    });
+    this.getSimulations();
+  }
+
+  getCompanies() {
+    this.restService.getData('companies/list')
+      .pipe(take(1))
+      .subscribe(
+        response => {
+          if (response && response.success) {
+            this.companies = response.data;
+          }
+        },
+        error => {
+          this.snackBarService.openSnackBar(error.error.message, '');
+        });
+  }
+
+  deleteSimulation(row) {
+    this.restService.deleteData(`simulations/delete/${row.sId}`)
+      .subscribe(
+        response => {
+          if (response && response.success) {
+            this.snackBarService.openSnackBar(response.message, '');
+            this.getSimulations();
+          }
+          this.loader = false;
+        }, error => {
+          this.snackBarService.openSnackBar(error.error.message, '');
+        }
+      );
+  }
+
+  simulate() {
+    this.loader = true;
+    this.restService.postData('simulations/start', this.simulateForm.value)
+      .pipe(take(1))
+      .subscribe(
+        response => {
+          if (response && response.success) {
+            this.getSimulations();
+          }
+          this.loader = false;
+        },
+        error => {
+          this.snackBarService.openSnackBar(error.error.message, '');
+          this.loader = false;
+
+        });
+  }
+
+  getSimulations() {
+    this.loader = true;
+    this.restService.getData('simulations/list')
+      .subscribe(
+        response => {
+          if (response && response.success) {
+            this.simulations = response.data;
+            this.dataSource = new MatTableDataSource(response.data);
+            this.dataSource.paginator = this.paginator;
+            this.dataSource.sort = this.sort;
+          }
+          this.loader = false;
+        },
+        error => {
+          this.snackBarService.openSnackBar(error.error.message, '');
+          this.loader = false;
+        });
+  }
+
+  updateSimulation(params) {
+    this.loader = true;
+    const payLoad = { sId: params.sId, status: (params.status) ? false : true }
+    this.restService.putData('simulations/update', payLoad)
+      .pipe(take(1))
+      .subscribe(
+        response => {
+          if (response && response.success) {
+            this.getSimulations();
+            this.snackBarService.openSnackBar(response.message, '');
+          }
+          this.loader = false;
+        },
+        error => {
+          this.snackBarService.openSnackBar(error.error.message, '');
+          this.loader = false;
+        });
+
+  }
+
+  isAlreadyStarted(id: string) {
+    if (this.simulations && this.simulations.find(el => el.companyId === id)) {
+      return true;
+    }
+    return false;
+  }
```

Comment:
calling **isAlreadyStarted** from html page for **[disabled]** property will really be very expensive in terms of performance, because we are iterating over simulations and matching current company id inside this function, and this function will be called for each companies and whenever user performs any action on any part of whole UI, this function will get executed again and again for each companies

I will suggest to maintain a flag for each company either from backend or else from frontend also once we get both response of companies and simulations we can iterate over it once and set a flag for each company and that can be used for **[disabled]** property

Reply:
this issue has been resolved and updated

Reply:
`updateCompanyStatus` function is using `this.companies`, but we are not sure like `this.companies` get api call will get finished before the function `updateCompanyStatus` will be called from ngOnInit right?
@vignesh-sdev

Reply:
this issue is fixed by calling companies list first and then simulations list on success response of company func.